File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    inc = &(&1 + 1)
    reducer = fn el, acc -> Map.update(acc, el, 1, inc) end

    sentence
    |> String.downcase()
    |> String.replace(~r/_/, " ")
    |> String.replace(~r/[:;!@#$%^&().,]/, "")
    |> String.replace(~r/\s\s+/, " ")
    |> String.split(" ")
    |> Enum.reduce(%{}, reducer)
  end
end

File: typescript/word-count/word-count.ts
export default class Words {
  public count(input: string): Map<string, number> {
    return input
      .trim()
      .replace(/\s+/, " ")
      .toLowerCase()
      .split(" ")
      .reduce((acc, el) => {
        acc.set(el, (acc.get(el) || 0) + 1);
        return acc;
      }, new Map<string, number>());
  }
}
